---
title: "read_and_clean_data_wecare_only"
output: html_document
---

started: Alexey Larionov, 2016  
last updated: Alexey Larionov, 03May2017

# Summary

Source data include output generated by the following wes pipeline versions:  
1) Wecare alignment to b37 of Oct2015  
2) NFE alignment to b37 of Oct2016
3) Joined wecare-nfe variant calling of Nov2016  
4) Variant filtering and annotation of Jan2017  

Also the source data include:  
- table for samples info, including wes-gwas correspondence and pass/fail info about samples in wes  
- several tables with phenotype data, obtained from DC at different times  
- table with cases where pathogenic BRCA1, BRCA2 or PALB2 variants were detected  

Exac and kgen include only variants that were biallelic in exac/ kgen respectively  
because GATK variant-to-table tool did not handle multiallelic sites correctly at the time of data generation.  

The script 
1) Reads data  
2) Does some clean-up (NA to missed values etc)  
3) Removes NFE data  

Output data: 343,824 vars x 512 cases  

# start_section

```{r start_section}

# Time stamp
Sys.time()


# Folders
library(knitr)
base_folder="/analysis/mtgroup_share/users/alexey/wecare_only_stat_05.17"
opts_knit$set(root.dir = base_folder)

scripts_folder <- paste(base_folder, "scripts", sep="/")
source_data_folder <- paste(base_folder, "source_data", sep="/")
interim_data_folder <- paste(base_folder, "interim_data", sep="/")
results_folder <- paste(base_folder, "results", sep="/")

```

# copy_source_ngs_data

Done only once - on Darwin cluster.  Hence FALSE in copying.  

```{r copy_source_ngs_data}

src_folder <- "/scratch/medgen/users/alexey/wecare_6_jan2017/wecare_nfe_nov2016_vqsr_shf_sma_ann_txt"
tgt_folder <- source_data_folder

prefix="wecare_nfe_nov2016_vqsr_shf_sma_ann"

gt_file <- paste(prefix,"GT_add.txt",sep="_")
gq_file <- paste(prefix,"GQ.txt",sep="_")
dp_file <- paste(prefix,"DP.txt",sep="_")
vv_file <- paste(prefix,"VV.txt",sep="_")
exac_file <- paste(prefix,"exac.txt",sep="_")
kgen_file <- paste(prefix,"kgen.txt",sep="_")

file.copy(
  paste(src_folder, gt_file, sep="/"),
  paste(tgt_folder, gt_file, sep="/"))

file.copy(
  paste(src_folder, gq_file, sep="/"),
  paste(tgt_folder, gq_file, sep="/"))

file.copy(
  paste(src_folder, dp_file, sep="/"),
  paste(tgt_folder, dp_file, sep="/"))

file.copy(
  paste(src_folder, vv_file, sep="/"),
  paste(tgt_folder, vv_file, sep="/"))

file.copy(
  paste(src_folder, exac_file, sep="/"),
  paste(tgt_folder, exac_file, sep="/"))

file.copy(
  paste(src_folder, kgen_file, sep="/"),
  paste(tgt_folder, kgen_file, sep="/"))

# Clean-up
rm(src_folder, tgt_folder)

```

# copy_source_phenotype_data

Done only once - on Darwin cluster.  Hence FALSE in copying.  

Files with phenotypes updates (Dec2016) and cases with BRCA1, BRCA2 and PALB2 (Dec 2016) were added manually.

```{r copy_source_phenotype_data}

src_folder <- "/scratch/medgen/users/alexey/wecare_stat_2_aug2016/source_data"
tgt_folder <- source_data_folder

covar_file <- "covar.txt"
samples_file <- "samples_ids.txt"
demographics_file <- "WECARE.Exome.DemographicVariables.txt"

file.copy(
  paste(src_folder, covar_file, sep="/"),
  paste(tgt_folder, covar_file, sep="/"))

file.copy(
  paste(src_folder, samples_file, sep="/"),
  paste(tgt_folder, samples_file, sep="/"))

file.copy(
  paste(src_folder, demographics_file, sep="/"),
  paste(tgt_folder, demographics_file, sep="/"))

# Clean-up
rm(src_folder, tgt_folder)

```

# read_data

```{r read_data}

gt.df <- read.table(
  paste(source_data_folder, gt_file, sep="/"), 
  header=TRUE, row.names=1, sep="\t", quote="")

gq.df <- read.table(
  paste(source_data_folder, gq_file, sep="/"), 
  header=TRUE, row.names=1, sep="\t", quote="")

dp.df <- read.table(
  paste(source_data_folder, dp_file, sep="/"), 
  header=TRUE, row.names=1, sep="\t", quote="")

covar.df <- read.table(
  paste(source_data_folder, covar_file, sep="/"), 
  sep="\t", header=TRUE, quote="")

samples.df <- read.table(
  paste(source_data_folder, samples_file, sep="/"), 
  sep="\t", header=TRUE, quote="")

demographics.df <- read.table(
  paste(source_data_folder, demographics_file, sep="/"), 
  sep="\t", header=TRUE, quote="")

vv.df <- read.table(
  paste(source_data_folder, vv_file, sep="/"), 
  header=TRUE, sep="\t", quote="")

kgen.df <- read.table(
  paste(source_data_folder, kgen_file, sep="/"), 
  header=TRUE, sep="\t", quote="")

exac.df <- read.table(
  paste(source_data_folder, exac_file, sep="/"), 
  header=TRUE, sep="\t", quote="")

phenotypes_update_file <- "phenotypes_update_06Dec2016.txt"
phenotypes_update.df <- read.table(
  paste(source_data_folder, phenotypes_update_file, sep="/"), 
  header=TRUE, sep="\t", quote="")

BRCA1_BRCA2_PALB2_cases_file <- "cases_with_BRCA1_BRCA2_PALB2_pathogenic_variants.txt"
BRCA1_BRCA2_PALB2_cases.df <- read.table(
  paste(source_data_folder, BRCA1_BRCA2_PALB2_cases_file, sep="/"), 
  header=TRUE, sep="\t", quote="")

# Update rownames, when necessary
rownames(vv.df) <- vv.df[,1]
rownames(kgen.df) <- kgen.df[,1]
rownames(exac.df) <- exac.df[,1]
rownames(BRCA1_BRCA2_PALB2_cases.df) <- BRCA1_BRCA2_PALB2_cases.df[,1]

# Update colnames, when necessary
colnames(gt.df) <- sub(".GT", "", colnames(gt.df))
colnames(gq.df) <- sub(".GQ", "", colnames(gq.df))
colnames(dp.df) <- sub(".DP", "", colnames(dp.df))

# Clean-up
rm(source_data_folder, prefix, covar_file, demographics_file, samples_file, vv_file, gt_file, gq_file, dp_file, kgen_file, exac_file, phenotypes_update_file, BRCA1_BRCA2_PALB2_cases_file)

```

# check_data

```{r check_data}

dim(gt.df)
str(gt.df, list.len=5)
gt.df[1:5,1:5]

dim(gq.df)
str(gq.df, list.len=5)
gq.df[1:5,1:5]

dim(dp.df)
str(dp.df, list.len=5)
dp.df[1:5,1:5]

dim(covar.df)
str(covar.df)
covar.df[1:5,1:5]

dim(samples.df)
str(samples.df)
samples.df[1:5,]

dim(demographics.df)
str(demographics.df)
demographics.df[1:5,1:5]

dim(phenotypes_update.df)
str(phenotypes_update.df)
phenotypes_update.df[1:5,1:5]

dim(BRCA1_BRCA2_PALB2_cases.df)
str(BRCA1_BRCA2_PALB2_cases.df)
BRCA1_BRCA2_PALB2_cases.df[1:5,1:5]

dim(vv.df)
str(vv.df)
vv.df[1:5,1:5]

dim(kgen.df)
str(kgen.df)
kgen.df[1:5,1:5]

dim(exac.df)
str(exac.df)
exac.df[1:5,1:5]

```

# convert_data_frames_to_matrices

```{r convert_data_frames_to_matrices}

gt.mx <- as.matrix(gt.df)
gq.mx <- as.matrix(gq.df)
dp.mx <- as.matrix(dp.df)

dim(gt.mx)
class(gt.mx)
gt.mx[1:5,1:5]

dim(gq.mx)
class(gq.mx)
gq.mx[1:5,1:5]

dim(dp.mx)
class(dp.mx)
dp.mx[1:5,1:5]

rm(gt.df, gq.df, dp.df)

```

# check_consistence_of_rownames_and_colnames

```{r check_consistence_of_rownames_and_colnames}

# rownames
sum(rownames(gt.mx) != rownames(gq.mx))
sum(rownames(gt.mx) != rownames(dp.mx))
sum(rownames(gt.mx) != rownames(vv.df))
sum(rownames(gt.mx) != rownames(kgen.df))
sum(rownames(gt.mx) != rownames(exac.df))

# colnames
sum(colnames(gt.mx) != colnames(gq.mx))
sum(colnames(gt.mx) != colnames(dp.mx))

```

# convert_blanks_to_NAs

```{r convert_blanks_to_NAs}

NA -> vv.df[vv.df$Existing_variation == "", "Existing_variation"] # no blanks in other fields
NA -> covar.df[covar.df == ""] # 1 case in chemo_hormone (row 84)
NA -> demographics.df[demographics.df == ""] # 2 cases in registry and one in chemo_hormone
NA -> BRCA1_BRCA2_PALB2_cases.df[BRCA1_BRCA2_PALB2_cases.df == ""] # 8 cases in notes

# No blanks in other tables

```

# reformat_phenotypes_update

```{r reformat_phenotypes_update}

# Converst factors to vectors (to symplify comparisons)
str(phenotypes_update.df)
as.character(phenotypes_update.df$gwas_id) -> phenotypes_update.df$gwas_id
as.character(phenotypes_update.df$wes_id) -> phenotypes_update.df$wes_id
trimws(as.character(phenotypes_update.df$hist_cat)) -> phenotypes_update.df$hist_cat
trimws(as.character(phenotypes_update.df$chemo_cat)) -> phenotypes_update.df$chemo_cat

as.character(phenotypes_update.df$er1) -> phenotypes_update.df$er1
as.character(phenotypes_update.df$pr1) -> phenotypes_update.df$pr1
as.character(phenotypes_update.df$hormone) -> phenotypes_update.df$hormone

NA -> phenotypes_update.df[phenotypes_update.df$er1 == "missing", "er1"]
NA -> phenotypes_update.df[phenotypes_update.df$pr1 == "missing", "pr1"]
NA -> phenotypes_update.df[phenotypes_update.df$hormone == "missing", "hormone"]

as.numeric(phenotypes_update.df$er1) -> phenotypes_update.df$er1
as.numeric(phenotypes_update.df$pr1) -> phenotypes_update.df$pr1
as.numeric(phenotypes_update.df$hormone) -> phenotypes_update.df$hormone

str(phenotypes_update.df)

```

# select_wecare_only

```{r}

wecare_cases <- grepl("^P",colnames(gq.mx))
sum(wecare_cases) # 512

dp.mx <- dp.mx[,wecare_cases]
gq.mx <- gq.mx[,wecare_cases]
gt.mx <- gt.mx[,wecare_cases]

rm(wecare_cases)

```

# data_summary

```{r data_summary}

dim(gt.mx)
class(gt.mx)
gt.mx[1:5,1:5]

dim(gq.mx)
class(gq.mx)
gq.mx[1:5,1:5]

dim(dp.mx)
class(dp.mx)
dp.mx[1:5,1:5]

dim(covar.df)
str(covar.df)
covar.df[1:5,1:5]

dim(samples.df)
str(samples.df)
samples.df[1:5,]

dim(demographics.df)
str(demographics.df)
demographics.df[1:5,1:5]

dim(phenotypes_update.df)
str(phenotypes_update.df)
phenotypes_update.df[1:5,1:5]

dim(BRCA1_BRCA2_PALB2_cases.df)
str(BRCA1_BRCA2_PALB2_cases.df)
BRCA1_BRCA2_PALB2_cases.df[1:5,1:5]

dim(vv.df)
str(vv.df)
vv.df[1:5,1:5]

dim(kgen.df)
str(kgen.df)
kgen.df[1:5,1:5]

dim(exac.df)
str(exac.df)
exac.df[1:5,1:5]

# Check consistence of rownames and colnames

sum(rownames(gt.mx) != rownames(gq.mx))
sum(rownames(gt.mx) != rownames(dp.mx))
sum(rownames(gt.mx) != rownames(vv.df))
sum(rownames(gt.mx) != rownames(kgen.df))
sum(rownames(gt.mx) != rownames(exac.df))

sum(colnames(gt.mx) != colnames(gq.mx))
sum(colnames(gt.mx) != colnames(dp.mx))

```

# save_data

```{r save_data}

save.image(paste(interim_data_folder, "r01_read_and_clean_data_wecare_only.RData", sep="/"))

```

# final_section

```{r final_section}

ls()
sessionInfo()
Sys.time()

```