---
title: "add_kgen50_to_wecare"
output: html_document
---

started: Alexey Larionov, 31Jan2017  
last updated: Alexey Larionov, 03May2017

# Summary

Adds kgen50 data (50 cases of different continental ancestory) to wecare-only data.  

Merged data will be used to generate ethnic-relevant set of eigenvectors  
to verify the ethnisity of the studied samples.  

The kgen50 data were generated in wecare-stat-Jan2017 analysis.  

Wecare data include output generated by the following wes pipeline versions:  
1) Alignment to b37 of Oct2015  
2) Variant calling of Nov2016  
3) Variant filtering and annotation of Jan2017  
Then processd by r01-r05 scripts in this folder.  

Input wecare data: 246,232 vars x 480 cases (245 UBC and 235 CBC)  
Input kgen50 data: 103,045 vars x 760 cases (512 wecare + 198 nfe + 50 kgen)  

Output kgen50-wecare data: 82,799 vars x 530 cases (480 wecare + 50 kgen)  

# start_section

```{r start_section}

# Time stamp
Sys.time()

# Folders
base_folder="/analysis/mtgroup_share/users/alexey/wecare_only_stat_05.17"
setwd(base_folder)
scripts_folder <- "scripts"
source_data_folder <- "source_data"
interim_data_folder <- "interim_data"
results_folder <- "results"

```

# read_data

**wecaare-only data**:  
246,232 vars x 480 cases (245 UBC and 235 CBC)  

**wecare-nfe-kgen50 data**  
Overlap of wecare-nfe and kgen50: 103,045 vars x 764 cases  

Variants:  
103,045 variants (present in each: wecare-nfe and kgen50)  

Samples:  
760 = 512 wecare + 198 nfe + 50 kgen  
and 4 additional columns: Chrom, Pos, kgen-id, wecare-nfe-id  

```{r read_data}

# load wecare-only data
load(paste(interim_data_folder, "r05_calculate_egenvectors_wecare_only.RData", sep="/"))

# Load kgen data

kgen50_cases.df <- read.table(paste(source_data_folder, "kgen_selected_50_samples.txt", sep="/"))
dim(kgen50_cases.df)

kgen50_vcf.df <- read.table(
  paste(source_data_folder, "wecare_nfe_kgen50_filt_vcf.txt", sep="/"), 
  header=TRUE, sep="\t", quote="")
dim(kgen50_vcf.df)

kgen50_gt.df <- read.table(
  paste(source_data_folder, "wecare_nfe_kgen50_filt_gt.txt", sep="/"), 
  header=TRUE, sep="\t", quote="")
dim(kgen50_gt.df)

# wecare-nfe-kgen50 gq and dp are not used because they are NA for kgen50 samples

# remove unnecessary data
rm(exac.df, kgen.df)

```

# update_kgen50_rownames

SplitVarId is the variants id from wecare-only  

```{r update_kgen50_rownames}

# Set rownames to SplitVarId (the VarId from wecare-nfe)
rownames(kgen50_vcf.df) <- as.vector(kgen50_vcf.df$SplitVarID)
rownames(kgen50_gt.df) <- as.vector(kgen50_gt.df$SplitVarID)

# Check the consistency of rows
sum(rownames(kgen50_vcf.df) != rownames(kgen50_gt.df))

```

# check_rename_and_select_kgen50_cases

Initial cases: 
760 = 512 wecare + 198 nfe + 50 kgen  

Selected cases:  
50 kgen  

```{r check_rename_and_select_kgen50_cases}

# Update kgen50_cases table
colnames(kgen50_cases.df) <- c("case", "subpopulation", "population", "gender")
kgen50_cases.df$case <- as.vector(kgen50_cases.df$case)
kgen50_cases.df$subpopulation <- as.vector(kgen50_cases.df$subpopulation)
kgen50_cases.df$population <- as.vector(kgen50_cases.df$population)
kgen50_cases.df$gender <- as.vector(kgen50_cases.df$gender)
str(kgen50_cases.df)
rownames(kgen50_cases.df) <- kgen50_cases.df$case

# Check counts of cases
kgen50_cases <- grep(".variant2.GT", colnames(kgen50_gt.df), value=TRUE)
length(kgen50_cases) # 50

wecare_nfe_cases <- grep(".variant.GT", colnames(kgen50_gt.df), value=TRUE)
length(wecare_nfe_cases) # 710

wecare_cases <- grep("^P", wecare_nfe_cases, value=TRUE)
length(wecare_cases) # 512

nfe_cases <- grep("^[H,N]", wecare_nfe_cases, value=TRUE)
length(nfe_cases) # 198

# Check kgen50 and nfe overlap
kgen50_cases <- sub(".variant2.GT","",kgen50_cases)
nfe_cases <- sub(".variant.GT","",nfe_cases)

sum(kgen50_cases %in% nfe_cases) # 9
# One kgen50 EUR (HG01704) was removed from our nfe set
# because of some odd QC metrics during our internal re-alignment

# Rename cases in kgen50 gt table
cases_raw <- colnames(kgen50_gt.df)
cases_renamed <- sub(".variant2.GT", "", cases_raw)
cases_renamed <- sub(".variant.GT", "_nfe", cases_renamed)
cases_renamed -> colnames(kgen50_gt.df)

# Keep kgen50 cases only
# (no NA data in genotypes - as expected)
selected_cases <- kgen50_cases.df$case
kgen50_gt.df <- kgen50_gt.df[,selected_cases]
sum(is.na(kgen50_gt.df)) # 0

# Clean-up
rm(kgen50_cases, wecare_nfe_cases, wecare_cases, nfe_cases, cases_raw, cases_renamed, selected_cases)

```

# remove_multiallelic_variants_in_kgen50

Remove 337 variants: 103,045 -> 102,708  

```{r remove_multiallelic_variants_in_kgen50}

ma <- grepl(",", as.vector(kgen50_vcf.df$ALT))

sum(ma) # 337

kgen50_gt.df <- kgen50_gt.df[!ma,]
kgen50_vcf.df <- kgen50_vcf.df[!ma,]

dim(kgen50_gt.df)
dim(kgen50_vcf.df)

rm(ma)

```

# convert_kgen50_gt_data_frames_to_matrix

```{r convert_kgen50_gt_data_frames_to_matrix}

class(kgen50_gt.df)

kgen50_gt.mx <- as.matrix(kgen50_gt.df)

class(kgen50_gt.mx)
str(kgen50_gt.mx)

rm(kgen50_gt.df)

```

# recode_kgen50_gt_to_additive

All phased genotypes, no missed data

```{r recode_kgen50_gt_to_additive}

# --- Prepare genotypes vectors --- #

# Ref and alt
ref <- as.vector(kgen50_vcf.df$REF)
alt <- as.vector(kgen50_vcf.df$ALT)
ref[1:5]
alt[1:5]

# Homozygous reference
ref_ref_phased <- paste(ref, ref, sep="|")
ref_ref_phased[1:5]

# Heterozygous
alt_ref_phased <- paste(alt, ref, sep="|")
ref_alt_phased <- paste(ref, alt, sep="|")
alt_ref_phased[1:5]
ref_alt_phased[1:5]

# Homozygous alt (biallelic sites only!)
alt_alt_phased <- paste(alt, alt, sep="|")
alt_alt_phased[1:5]

# --- Convert genotypes to additive --- #

# Homozygous reference
0 -> kgen50_gt.mx[kgen50_gt.mx == ref_ref_phased]

# Heterozygous
1 -> kgen50_gt.mx[kgen50_gt.mx == alt_ref_phased]
1 -> kgen50_gt.mx[kgen50_gt.mx == ref_alt_phased]

# Homozygous alt (biallelic sites only!)
2 -> kgen50_gt.mx[kgen50_gt.mx == alt_alt_phased]

# Check the re-coding result
summary(as.factor(kgen50_gt.mx))

# Set rownames and colnames
rownames_gt <- rownames(kgen50_gt.mx)
colnames_gt <- colnames(kgen50_gt.mx)
kgen50_gt.mx <- matrix(as.numeric(kgen50_gt.mx), nrow=nrow(kgen50_gt.mx))
rownames_gt -> rownames(kgen50_gt.mx)
colnames_gt -> colnames(kgen50_gt.mx)

# Make sure the matrix is numeric
min(kgen50_gt.mx, na.rm=TRUE)
max(kgen50_gt.mx, na.rm=TRUE)

# Check dimentions of gt
dim(kgen50_gt.mx)

# Clean-up
rm(ref, alt, ref_ref_phased, alt_ref_phased, ref_alt_phased, alt_alt_phased, rownames_gt, colnames_gt)

```

# remove_variants_with_the_uniform_genotypes_accross_all_kgen50_samples

Remove 1,086 variants: 102,708 -> 101,622   

```{r remove_variants_with_the_uniform_genotypes_accross_all_kgen50_samples}

# Function to detect uniform numeric vector
uniform_vector.fnc <- function(x){
  if(min(x, na.rm=TRUE) == max(x, na.rm=TRUE)){return(TRUE)} else {return(FALSE)}}

# Variants with uniform genotypes accross all samples 
uniform_genotypes <- apply(kgen50_gt.mx, 1, uniform_vector.fnc)
summary(uniform_genotypes)
sum(uniform_genotypes)

# Remove variants with uniform genotypes accross all samples
kgen50_gt.mx <- kgen50_gt.mx[!uniform_genotypes,]
kgen50_vcf.df <- kgen50_vcf.df[!uniform_genotypes,]

# Check new size of data
dim(kgen50_gt.mx)
dim(kgen50_vcf.df)

# Clean-up
rm(uniform_vector.fnc, uniform_genotypes)

```

# estimate_number_of_low_frequency_variants

Apparent excess of group-A variants (AF>95%) over the group-B (AF<5%) may reflects  
preferential selection of variants with rare allele in reference genome: why would it happen?  

```{r estimate_number_of_low_frequency_variants}

sum(as.vector(kgen50_vcf.df$AF) > 0.95) # rare variants with rare allele in reference genome (group A)
# 29,563

sum(as.vector(kgen50_vcf.df$AF) < 0.05) # rare variants with common allele in reference genome (group B)
# 19,462

```

# merge_kgen50_with_wicare-only

Select overlapped variants: 101,622 -> 82,799  
Add cases: 530 = 480 wecare + 50 nfe  

```{r merge_kgen50_with_wicare-nfe}

# Select kgen50 variants overlapping with wecare-only

kgen50_split_var_ids <- as.vector(kgen50_vcf.df$SplitVarID)
length(kgen50_split_var_ids) # 101,622

wecare_only_split_var_ids <- as.vector(variants.df$SplitVarID)
length(wecare_only_split_var_ids) # 246,232

overlap_split_var_ids <- intersect(kgen50_split_var_ids, wecare_only_split_var_ids)
length(overlap_split_var_ids) # 82,799

# Make merged genotypes table
wecare_kgen50_gt.mx <- cbind(genotypes.mx[overlap_split_var_ids,], kgen50_gt.mx[overlap_split_var_ids,])
dim(wecare_kgen50_gt.mx) # 82,799 x 530

# Subset variants and annotations
wecare_kgen50_vars.df <- variants.df[overlap_split_var_ids,]
dim(wecare_kgen50_vars.df)

# Prepare group labels for the merged cases
wecare_kgen50_groups.df <- cbind(case=colnames(wecare_kgen50_gt.mx), group="")
wecare_kgen50_groups.df[,"case"] -> rownames(wecare_kgen50_groups.df)

wecare_groups <- phenotypes.df$cc
"CBC" -> wecare_groups[wecare_groups==1]
"UBC" -> wecare_groups[wecare_groups==0]
kgen50_groups <- as.vector(kgen50_cases.df$population)

wecare_kgen50_groups.df[,"group"] <- c(wecare_groups, kgen50_groups)

# Clean-up
rm(kgen50_vcf.df, kgen50_gt.mx, kgen50_split_var_ids, wecare_only_split_var_ids, 
   overlap_split_var_ids, genotypes.mx, phenotypes.df, variants.df, 
   wecare_groups, kgen50_groups, kgen50_cases.df)

```

# data_summary

```{r data_summary}

ls()

dim(wecare_kgen50_gt.mx)
class(wecare_kgen50_gt.mx)
wecare_kgen50_gt.mx[1:5, 1:5]

dim(wecare_kgen50_groups.df)
str(wecare_kgen50_groups.df)
wecare_kgen50_groups.df[1:5, ]

dim(wecare_kgen50_vars.df)
str(wecare_kgen50_vars.df)
wecare_kgen50_vars.df[1:5,1:5]

# Check consistence of rownames and colnames
sum(rownames(wecare_kgen50_gt.mx) != rownames(wecare_kgen50_vars.df))
sum(colnames(wecare_kgen50_gt.mx) != rownames(wecare_kgen50_groups.df))

```

# save_data

```{r save_data}

save.image(paste(interim_data_folder, "r06a_add_kgen50_to_wecare.RData", sep="/"))

```

# final_section

```{r final_section}

ls()
sessionInfo()
Sys.time()

```